clientconn.go

```diff
@@ -840,6 +840,73 @@ func (ac *addrConn) connect() error {
 	ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
 	ac.mu.Unlock()
 
+	common := func() bool {
+		ac.mu.Lock()
+		if ac.state == connectivity.Shutdown {
+			ac.mu.Unlock()
+			return true
+		}
+
+		ac.state = connectivity.TransientFailure
+		ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
+		ac.cc.resolveNow(resolver.ResolveNowOption{})
+		ac.curAddr = resolver.Address{}
+		ac.mu.Unlock()
+
+		if err := ac.resetTransport(); err != nil {
+			ac.mu.Lock()
+			ac.printf("transport exiting: %v", err)
+			ac.mu.Unlock()
+			grpclog.Warningf("grpc: addrConn.transportMonitor exits due to: %v", err)
+			if err != errConnClosing {
+				// Keep this ac in cc.conns, to get the reason it's torn down.
+				ac.tearDown(err)
+			}
+			return true
+		}
+		return false
+	}
+
+	var timer *time.Timer
+	var cdeadline <-chan time.Time
+
+	onDeadline := func() bool {
+		ac.mu.Lock()
+		if ac.backoffDeadline.IsZero() {
+			ac.mu.Unlock()
+			return true
+		}
+		ac.mu.Unlock()
+		err := ac.transport.Close()
+		if err != nil {
+			grpclog.Error(err)
+		}
+
+		return common()
+	}
+
+	onGoAway := func() bool {
+		if timer != nil {
+			timer.Stop()
+		}
+		ac.adjustParams(ac.transport.GetGoAwayReason())
+		return common()
+	}
+
+	onError := func() bool {
```

Comment:
In the current implementation, the closures aren't passed into the transport. They're only invoked up at the connect() level. So, once a transport is reset (as part of onError or onGoAway), there is no chance for the closure to be called again on an old transport since the for loop is always concerned with the most recent transport.